src/AutoSelect.ml

```diff
@@ -0,0 +1,209 @@
+
+let disableIntervalContextsInRecursiveFunctions () =
+  ResettableLazy.force functionCallMaps |> fun (x,_,_) -> x |> FunctionCallMap.iter (fun f set ->
+    (*detect direct recursion and recursion with one indirection*)
+    if FunctionSet.mem f set || (not @@ FunctionSet.disjoint (calledFunctions f) (callingFunctions f)) then (
+      print_endline ("function " ^ (f.vname) ^" is recursive, disable interval analysis");
+      f.vattr <- addAttributes (f.vattr) [Attr ("goblint_context",[AStr "base.no-interval"])];
+    )
+  );;
+
+(*If only one Thread is used in the program, we can disable most thread analyses*)
+(*The exceptions are analyses that are depended on by others: base -> mutex -> mutexEvents, access*)
+(*TODO escape is also still enabled, because I do not know if the analysis gets more imprecise if not*)
+
+let toJsonArray list = "[\"" ^ (String.concat "\",\"" list) ^ "\"]";;
+let notNeccessaryThreadAnalyses = ["deadlock"; "maylocks"; "symb_locks"; "thread"; "threadflag"; "threadid"; "threadJoins"; "threadreturn"];;
+
+let reduceThreadAnalyses () = 
+  (*TODO also consider dynamic calls!?*)
+  let hasThreadCreate () = 
+    ResettableLazy.force functionCallMaps 
+    |> fun (_,x,_) -> x  (*every function that is called*)
+    |> FunctionCallMap.exists
+      (fun var callers ->
+        let desc = LibraryFunctions.find var in
+          match (functionArgs var) with
+            | None -> false;
+            | Some args -> 
+              match desc.special args with
+              | ThreadCreate _ -> 
+                print_endline @@ "thread created in " ^ var.vname ^ ", called by:"; 
+                FunctionSet.iter ( fun c -> print_endline @@ "  " ^ c.vname) callers;
+                true;
+              | _ -> false; 
+      ) 
+  in
+
+  (*TODO is there a way to specify only the thread analyses to keep?  *)
+  if not @@ hasThreadCreate () then (
+      print_endline @@ "no thread creation -> disabeling thread analyses \"" ^ (String.concat ", " notNeccessaryThreadAnalyses) ^ "\"";
+      get_string_list "ana.activated" 
+      |> List.filter (fun l -> not (List.mem l notNeccessaryThreadAnalyses))
+      |> toJsonArray
+      |> set_auto "ana.activated"
+
+    )
+ ;;
+
+let focusOnSpecification () = 
+  match Svcomp.Specification.of_option () with 
+    | UnreachCall s -> () (*TODO?*)
+    | NoDataRace -> (*enable all thread analyses*)
+      print_endline @@ "Specification: NoDataRace -> enabeling thread analyses \"" ^ (String.concat ", " notNeccessaryThreadAnalyses) ^ "\"";
+      get_string_list "ana.activated" 
+      |> List.filter (fun l -> not (List.mem l notNeccessaryThreadAnalyses)) (*remove duplicates*)
+      |> (@) notNeccessaryThreadAnalyses
+      |> toJsonArray
+      |> set_auto "ana.activated";
+    | NoOverflow -> (*We focus on integer analysis*)
+      set_bool "ana.int.def_exc" true;
+      set_bool "ana.int.enums" true;
+      set_bool "ana.int.interval" true;;
```

Comment:
Instead of hard-coding the options to change for each specification, ultimately the nicer approach would be something like what CPAchecker does: have a conf file for each specification and just load them.

Reply:
That is a good suggestion. Because I do not know if this is in any way final, I would not implement it yet.

Reply:
I would suggest to leave it like this for this PR.